### PR TITLE
IGNITE-955 Local listener in continuous queries should not be mandatory

### PR DESCRIPTION
**Pull request by ololo3000**

### CODE REVIEW

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java [hunk 1]
```diff
@@ -105,6 +111,12 @@
     /** */
     private static final String NO_CACHE_IGNITE_INSTANCE_NAME = "noCacheGrid";
 
+    /**
+     * Remote filter latch. Executed on remote nodes, so must be transient to not get serialization exception.
```
Comment (1vanan) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java]: Remote filter latch is executed on remote nodes, so must be transient to **do** not get serialization exception.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheProxyImpl.java [hunk 2]
```diff
@@ -553,8 +553,11 @@ private ClusterGroup projection(boolean loc) {
         if (qry instanceof ContinuousQuery) {
             ContinuousQuery<K, V> qry0 = (ContinuousQuery<K, V>)qry;
 
-            if (qry0.getLocalListener() == null)
-                throw new IgniteException("Mandatory local listener is not set for the query: " + qry);
+            if (qry0.getLocalListener() == null &&
+                qry0.getRemoteFilterFactory() == null && qry0.getRemoteFilter() == null) {
```
Comment (1vanan) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheProxyImpl.java]: Is this situation possible: qry0.getRemoteFilter() is not null, but qry0.getRemoteFilterFactory() returns null?

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java [hunk 3]
```diff
@@ -1205,6 +1215,171 @@ public void testExpired() throws Exception {
...
+                filtered.set(true);
+
+                return bypassFilter;
+            }));
+
+        QueryCursor<Cache.Entry<Integer, Integer>> qryCursor = grid(0).cache(DEFAULT_CACHE_NAME).query(qry);
+
+        grid(0).<Integer, Integer>cache(DEFAULT_CACHE_NAME).put(1, 1);
+
+        assertTrue(GridTestUtils.waitForCondition(() -> filtered.get(), 100));
```
Comment (1vanan) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java]: Please, move the timeout to separate variable here and in another waitForCondition places.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java [hunk 4]
```diff
@@ -1205,6 +1215,171 @@ public void testExpired() throws Exception {
...
+        ContinuousQuery<Integer, Integer> qry = new ContinuousQuery<>();
+
+        qry.setRemoteFilterFactory(
+            FactoryBuilder.factoryOf((CacheEntryEventSerializableFilter<Integer, Integer>)evt -> {
+                filtered.set(true);
+
+                return bypassFilter;
+            }));
+
+        QueryCursor<Cache.Entry<Integer, Integer>> qryCursor = grid(0).cache(DEFAULT_CACHE_NAME).query(qry);
```
Comment (1vanan) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java]: Move it to try block.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java [hunk 5]
```diff
@@ -1205,6 +1215,171 @@ public void testExpired() throws Exception {
...
+
+        ContinuousQuery<Integer, Integer> qry = new ContinuousQuery<>();
+
+        qry.setRemoteFilter(evt -> {
+                filtered.set(true);
+
+                return bypassFilter;
+            });
+
+        QueryCursor<Cache.Entry<Integer, Integer>> qryCursor = grid(0).cache(DEFAULT_CACHE_NAME).query(qry);
```
Comment (1vanan) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java]: Move it to try block.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java [hunk 6]
```diff
@@ -1205,6 +1215,171 @@ public void testExpired() throws Exception {
...
+        qry.setRemoteTransformerFactory(FactoryBuilder.factoryOf(
+            new IgniteClosure<CacheEntryEvent<? extends Integer, ? extends Integer>, Integer>() {
+                @Override public Integer apply(CacheEntryEvent<? extends Integer, ? extends Integer> evt) {
+                    transformed.set(true);
+
+                    return evt.getValue();
+                }
+            }));
+
+        QueryCursor<Cache.Entry<Integer, Integer>> qryCursor = grid(0).cache(DEFAULT_CACHE_NAME).query(qry);
```
Comment (1vanan) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java]: Move it to try block.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java [hunk 7]
```diff
@@ -1205,6 +1215,171 @@ public void testExpired() throws Exception {
...
+        qry.setInitialQuery(new ScanQuery<>((k, v) -> k == 0));
+
+        qry.setRemoteFilterFactory(
+            FactoryBuilder.factoryOf((CacheEntryEventSerializableFilter<Integer, Integer>)evt -> {
+                filtered.set(true);
+
+                return bypassFilter;
+            }));
+
+        QueryCursor<Cache.Entry<Integer, Integer>> qryCursor = grid(0).cache(DEFAULT_CACHE_NAME).query(qry);
```
Comment (1vanan) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java]: Move it to try block.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java [hunk 8]
```diff
@@ -1205,6 +1215,171 @@ public void testExpired() throws Exception {
...
+                filtered.set(true);
+
+                return bypassFilter;
+            });
+
+        QueryCursor<Cache.Entry<Integer, Integer>> qryCursor = grid(0).cache(DEFAULT_CACHE_NAME).query(qry);
+
+        grid(0).<Integer, Integer>cache(DEFAULT_CACHE_NAME).put(1, 1);
+
+        assertTrue(GridTestUtils.waitForCondition(() -> filtered.get(), 100));
```
Comment (1vanan) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java]: Move the timeout to the separate variable.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java [hunk 9]
```diff
@@ -1205,6 +1215,171 @@ public void testExpired() throws Exception {
...
+
+                    return evt.getValue();
+                }
+            }));
+
+        QueryCursor<Cache.Entry<Integer, Integer>> qryCursor = grid(0).cache(DEFAULT_CACHE_NAME).query(qry);
+
+        grid(0).<Integer, Integer>cache(DEFAULT_CACHE_NAME).put(1, 1);
+
+        assertTrue(GridTestUtils.waitForCondition(() -> filtered.get(), 100));
```
Comment (1vanan) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java]: Move the timeout to the separate variable.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java [hunk 10]
```diff
@@ -1205,6 +1215,171 @@ public void testExpired() throws Exception {
...
+
+        if (bypassFilter && gridCount() > 1)
+            assertTrue(GridTestUtils.waitForCondition(() -> transformed.get(), 100));
```
Comment (1vanan) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java]: Move the timeout to the separate variable.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java [hunk 11]
```diff
@@ -1205,6 +1215,171 @@ public void testExpired() throws Exception {
...
+
+        assertEquals(1, res.size());
+
+        assertEquals(0, res.get(0).getKey().intValue());
+
+        assertEquals(0, res.get(0).getValue().intValue());
+
+        grid(0).<Integer, Integer>cache(DEFAULT_CACHE_NAME).put(1, 1);
+
+        assertTrue(GridTestUtils.waitForCondition(() -> filtered.get(), 100));
```
Comment (1vanan) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java]: Move the timeout to the separate variable.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java [hunk 12]
```diff
@@ -30,9 +30,11 @@
 import java.util.concurrent.Callable;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
```
Comment (1vanan) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java]: Please, remove unnecessary import.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java [hunk 13]
```diff
@@ -105,6 +109,12 @@
     /** */
     private static final String NO_CACHE_IGNITE_INSTANCE_NAME = "noCacheGrid";
 
+    /** Tracks {@link CacheContinuousQueryHandler#filter(CacheContinuousQueryEvent) calls}. */
+    private static AtomicBoolean filtered = new AtomicBoolean(false);
```
Comment (1vanan) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java]: Make it final.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java [hunk 14]
```diff
@@ -105,6 +109,12 @@
...
+
+    /** Tracks {@link CacheContinuousQueryHandler#transform(IgniteClosure, CacheEntryEvent)}  calls}.*/
+    private static AtomicBoolean transformed = new AtomicBoolean(false);
```
Comment (1vanan) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/GridCacheContinuousQueryAbstractSelfTest.java]: Make it final.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheProxyImpl.java [hunk 15]
```diff
@@ -566,8 +569,10 @@ private ClusterGroup projection(boolean loc) {
         else {
             ContinuousQueryWithTransformer<K, V, ?> qry0 = (ContinuousQueryWithTransformer<K, V, ?>)qry;
 
-            if (qry0.getLocalListener() == null)
-                throw new IgniteException("Mandatory local transformed event listener is not set for the query: " + qry);
+            if (qry0.getLocalListener() == null && qry0.getRemoteFilterFactory() == null) {
```
Comment (1vanan) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheProxyImpl.java]: May be check on qry0.getRemoteFilter() is also necessary here?

Comment (ololo3000) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheProxyImpl.java]: ContinuousQueryWithTransformer doesn't provide this method.

**Pull request merged**